giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+        throw new Exception("There are " + coefficient.length +
+          " coefficients, but " + validColumnIndices.size() +
+          " valid columns in the regression");
+      }
+
+      double[] beta = mlr.estimateRegressionParameters();
+      for (int i = 0; i < coefficient.length; ++i) {
+        coefficient[i] = 0;
+      }
+      for (int i = 0; i < validColumnIndices.size(); ++i) {
+        coefficient[validColumnIndices.get(i)] = beta[i];
+      }
+      coefficient[5] = beta[validColumnIndices.size()];
+    }
+
+    /**
+     * Copies the values from a Vector to an array.
+     * @param source Source vector of values
+     * @param target Target array.
+     */
+    private static void copyVectorToArray(Vector<Double> source,
+                                          double[] target) {
+      for (int i = 0; i < source.size(); ++i) {
+        target[i] = source.get(i);
+      }
+    }
+
+    /**
+     * Copies values from a Vector of double[] to a double[][]. Takes into
+     * consideration the list of valid column indices.
+     * @param sourceValues Source Vector of double[]
+     * @param validColumnIndices Valid column indices
+     * @param xValues Target double[][] matrix.
+     */
+    private static void fillXMatrix(Vector<double[]> sourceValues,
+      Vector<Integer> validColumnIndices, double[][] xValues) {
+
+      for (int i = 0; i < sourceValues.size(); ++i) {
+        xValues[i] = new double[validColumnIndices.size() + 1];
+        for (int j = 0; j < validColumnIndices.size(); ++j) {
+          xValues[i][j] = sourceValues.get(i)[validColumnIndices.get(j)];
+        }
+        xValues[i][validColumnIndices.size()] = 1;
+      }
+    }
+
+    /**
+     * Utility function that checks whether two doubles are equals given an
+     * accuracy tolerance.
+     *
+     * @param val1 First value
+     * @param val2 Second value
+     * @return True if within a threshold
+     */
+    private static boolean equal(double val1, double val2) {
+      return Math.abs(val1 - val2) < 0.01;
+    }
+
+    /**
+     * Utility function that checks if two columns have linear dependence.
+     *
+     * @param values Matrix in the form of a Vector of double[] values.
+     * @param col1 First column index
+     * @param col2 Second column index
+     * @return True if there is linear dependence.
+     */
+    private static boolean isLinearDependence(Vector<double[]> values,
+                                              int col1, int col2) {
+
+      Vector<Double> factor = new Vector<>();
+      for (double[] value : values) {
+        double val1 = value[col1];
+        double val2 = value[col2];
+        if (equal(val1, 0)) {
+          if (equal(val2, 0)) {
+            continue;
+          } else {
+            return false;
+          }
+        }
+        if (equal(val2, 0)) {
+          return false;
+        }
+        factor.add(val1 / val2);
+      }
+
+      if (factor.size() < 2) {
```

Comment:
Not really. `factors.size()` is the number of non-zero elements in the two columns of the given matrix. For instance, the following can be the matrix:
0, 0
1, 2
3, 5
0, 0
Here `values.length` is 4, but `factor.size()` is 2. Basically, the loop above this line weed out the rows that are 0 in both columns from the dependency check.